File: Python/bucket_sort.py
import math

def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

def bucket_sort(arr):
    if len(arr) == 0:
        return arr

    max_val = max(arr)
    min_val = min(arr)
    bucket_count = len(arr)
    size = math.ceil((max_val - min_val + 1) / bucket_count)

    buckets = [[] for _ in range(bucket_count)]
    for num in arr:
        index = (num - min_val) // size
        buckets[index].append(num)

    for i in range(len(buckets)):
        buckets[i] = insertion_sort(buckets[i])

    result = []
    for bucket in buckets:
        result.extend(bucket)
    return result

arr = [3, 7, 4, 8, 6, 2, 1, 5, 0, -5, -2]
print(bucket_sort(arr))

File: C#/BucketSort.cs
using System;
using System.Collections.Generic;

class BucketSort
{
    public static void Sort(int[] arr)
    {
        int n = arr.Length;
        if (n <= 0) return;

        int maxVal = arr[0], minVal = arr[0];

        for (int i = 1; i < n; i++)
        {
            if (arr[i] > maxVal) maxVal = arr[i];
            if (arr[i] < minVal) minVal = arr[i];
        }

        int bucketCount = Math.Max(1, (maxVal - minVal) / n + 1);
        List<int>[] buckets = new List<int>[bucketCount];

        for (int i = 0; i < bucketCount; i++)
            buckets[i] = new List<int>();

        foreach (var num in arr)
        {
            int bucketIndex = (num - minVal) * (bucketCount - 1) / (maxVal - minVal + 1);
            buckets[bucketIndex].Add(num);
        }

        for (int i = 0; i < bucketCount; i++)
            buckets[i].Sort();

        int index = 0;
        for (int i = 0; i < bucketCount; i++)
            foreach (var num in buckets[i])
                arr[index++] = num;
    }

    public static void Main()
    {
        int[] arr = { 78, 17, 39, 26, 72, 94, 21, 40, 55, 69 };
        Sort(arr);
        Console.WriteLine(string.Join(" ", arr));
    }
}
